src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/UserDefinedImplicitConversions.cs

```diff
@@ -278,7 +278,8 @@ private static void ComputeUserDefinedImplicitConversionTypeSet(TypeSymbol s, Ty
 
                         u.Add(UserDefinedConversionAnalysis.Normal(op, fromConversion, toConversion, convertsFrom, convertsTo));
                     }
-                    else if ((object)source != null && (object)target != null && source.IsNullableType() && convertsFrom.IsNonNullableValueType() && target.CanBeAssignedNull())
+                    else if ((object)source != null && source.IsNullableType() && convertsFrom.IsNonNullableValueType() &&
+                        ((object)target != null ? target.CanBeAssignedNull() : convertsTo.IsReferenceType))
```

Comment:
@AlekseyTs made two comments to me (by phone) regarding this.

1. The enclosing method should clearly document that `allowAnyTarget==true` (or equivalently, `(object)target == null`) occurs only on the controlling expression of a switch statement
2. It isn't clear that this code will properly lift a user-defined implicit conversion operator that converts from a struct type `X` to `int?`, when the controlling expression of a switch statement is of type `X?`. That should at least be tested, and the code adjusted so that such a test passes.